### PR TITLE
updated /deliveryservices/{{ID}}/servers and /deliveryservices/{{ID}}…

### LINKED ISSUE #4614: Update /deliveryservices/{{ID}}/servers to use data from new interface tables
## I'm submitting a ...
- new feature / enhancement request

## Description
The `/deliveryservices/{{ID}}/servers` endpoint need to be updated to utilize data from the new tables added by whatever PR resolves #4611.

This depends on #4611

## Traffic Control components affected ...
- Documentation
- Traffic Ops
- Traffic Ops client (Go)

## Acceptance Criteria
* `/deliveryservices/{{ID}}/servers` in API 1.x:
	* Uses the tables to provide the right legacy output for the `GET` method (should be unchanged)
* `/deliveryservices/{{ID}}/servers` in API 2.x:
	* Provides the proper representations of cache servers as outlined in the blueprint for the `GET` method
	* New behavior adheres to the API guidelines (where applicable)
	* All behavior changes have been documented
* The Traffic Ops Go client has been updated (v2 only) to return the new structures
	* Any associated unit and/or client/API integration tests have been updated to
	  test the new structure and all pass.

See the [Additional Server Interfaces blueprint](https://github.com/apache/trafficcontrol/blob/master/blueprints/multi-interface-servers.md#deliveryservicesidservers) for details.

### LINKED ISSUE #4615: Update /deliveryservices/{{ID}}/servers/eligible to use data from new interfaces tables
## I'm submitting a ...
- new feature / enhancement request

## Description
The `/deliveryservices/{{ID}}/servers/eligible` endpoint need to be updated to utilize data from the new tables added by whatever PR resolves #4611.

This depends on #4611

## Traffic Control components affected ...
- Documentation
- Traffic Ops
- Traffic Ops client (Go)

## Acceptance Criteria
* `/deliveryservices/{{ID}}/servers/eligible` in API 2.x:
	* Uses the tables to provide the right legacy output for the `GET` method (should be unchanged)
* `/deliveryservices/{{ID}}/servers/eligible` in API 3.x:
	* Provides the proper representations of cache servers as outlined in the blueprint for the `GET` method
	* New behavior adheres to the API guidelines (where applicable)
	* All behavior changes have been documented
* The Traffic Ops Go client has been updated (v3 only) to return the new structures
	* Any associated unit and/or client/API integration tests have been updated to
	  test the new structure and all pass.

See the [Additional Server Interfaces blueprint](https://github.com/apache/trafficcontrol/blob/master/blueprints/multi-interface-servers.md#deliveryservicesidserverseligible) for details.

### PR DESCRIPTION
**Pull request by mattjackson220**
…/servers/eligible to use multiple interfaces

## What does this PR (Pull Request) do?

- [x] This PR fixes #4614  and fixes #4615  

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Ops

## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] This PR includes documentation
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: lib/go-tc/deliveryservice_servers.go [hunk 1]
```diff
@@ -114,3 +100,13 @@ type DSServer struct {
...
+
+type DSServerV11 struct {
+	DSServer
+	LegacyInterfaceDetails
+}
+
+type DSServerV30 struct {
+	DSServer
+	ServerInterfaces *[]ServerInterfaceInfo `json:"interfaces" db:"interfaces"`
+}
```
Comment (ocket8888) [inline lib/go-tc/deliveryservice_servers.go:112]: GoDoc on these?

File: lib/go-tc/servers.go [hunk 2]
```diff
@@ -267,6 +274,20 @@ type ServerDetail struct {
 	XMPPPasswd         *string           `json:"xmppPasswd" db:"xmpp_passwd"`
 }
 
+// LegacyInterfaceDetails is the details for interfaces on servers for API v1 and v2
```
Comment (ocket8888) [inline lib/go-tc/servers.go:277]: nit: GoDocs should end with a period.

File: lib/go-tc/servers.go [hunk 3]
```diff
@@ -286,3 +307,88 @@ type ServerQueueUpdate struct {
 	ServerID util.JSONIntStr `json:"serverId"`
 	Action   string          `json:"action"`
 }
+
+// ServerInterface is the data associated with an interface
```
Comment (ocket8888) [inline lib/go-tc/servers.go:311]: same as above RE: period

File: lib/go-tc/servers.go [hunk 4]
```diff
@@ -286,3 +307,88 @@ type ServerQueueUpdate struct {
...
+type ServerInterface struct {
+	MaxBandwidth int64  `json:"max_bandwidth" db:"max_bandwidth"`
+	Monitor      bool   `json:"monitor" db:"monitor"`
+	Mtu          int64  `json:"mtu" db:"mtu"`
+	Name         string `json:"name" db:"name"`
+	Server       int    `json:"server" db:"server"`
+}
+
+// ServerIpAddress is the data associated with an IP address
```
Comment (ocket8888) [inline lib/go-tc/servers.go:320]: same as above RE: period

File: lib/go-tc/servers.go [hunk 5]
```diff
@@ -286,3 +307,88 @@ type ServerQueueUpdate struct {
...
+type ServerIpAddress struct {
+	Address        string `json:"address" db:"address"`
+	Gateway        string `json:"gateway" db:"gateway"`
+	Interface      string `json:"interface,omitempty" db:"interface"`
+	Server         int    `json:"server,omitempty" db:"server"`
+	ServiceAddress bool   `json:"service_address" db:"service_address"`
+}
+
+// ServerInterfaceInfo is the combined data from both IP address information and interface information
```
Comment (ocket8888) [inline lib/go-tc/servers.go:329]: same as above RE: period

File: lib/go-tc/servers.go [hunk 6]
```diff
@@ -286,3 +307,88 @@ type ServerQueueUpdate struct {
...
+type ServerInterfaceInfo struct {
+	IpAddresses  []ServerIpAddress `json:"ipAddresses" db:"ipAddresses"`
+	MaxBandwidth int64             `json:"max_bandwidth" db:"max_bandwidth"`
+	Monitor      bool              `json:"monitor" db:"monitor"`
+	Mtu          int               `json:"mtu" db:"mtu"`
```
Comment (ocket8888) [inline lib/go-tc/servers.go:334]: MTU should be `uint` (or `uint64`) - can't ever be valid as a negative number

File: lib/go-tc/servers.go [hunk 7]
```diff
@@ -286,3 +307,88 @@ type ServerQueueUpdate struct {
...
+	Name         string            `json:"name" db:"name"`
+}
+
+// Value implements the driver.Valuer interface
+// marshals struct to json to pass back as a json.RawMessage
```
Comment (ocket8888) [inline lib/go-tc/servers.go:339]: nit: GoDocs should consist of complete sentences with proper capitalization and terminated by periods.

File: lib/go-tc/servers.go [hunk 8]
```diff
@@ -286,3 +307,88 @@ type ServerQueueUpdate struct {
...
+func (sii *ServerInterfaceInfo) Value() (driver.Value, error) {
+	b, err := json.Marshal(sii)
+	return b, err
+}
+
+// Scan implements the sql.Scanner interface
+// expects json.RawMessage and unmarshals to a deliveryservice struct
```
Comment (ocket8888) [inline lib/go-tc/servers.go:346]: Same as above RE: complete sentences

File: lib/go-tc/servers.go [hunk 9]
```diff
@@ -286,3 +307,88 @@ type ServerQueueUpdate struct {
...
+func (sii *ServerInterfaceInfo) Scan(src interface{}) error {
+	b, ok := src.([]byte)
+	if !ok {
+		return fmt.Errorf("expected deliveryservice in byte array form; got %T", src)
+	}
+
+	return json.Unmarshal([]byte(b), sii)
+}
+
+// ConvertInterfaceInfotoV11 converts ServerInterfaceInfo to LegacyInterfaceDetails for v1 and v2 compatibility
```
Comment (ocket8888) [inline lib/go-tc/servers.go:356]: Same as above RE: period.

File: lib/go-tc/deliveryservice_servers.go [hunk 10]
```diff
@@ -114,3 +100,13 @@ type DSServer struct {
 	ServerCapabilities          []string             `json:"-" db:"server_capabilities"`
 	DeliveryServiceCapabilities []string             `json:"-" db:"deliveryservice_capabilities"`
 }
+
+type DSServerV11 struct {
+	DSServer
+	LegacyInterfaceDetails
+}
+
+type DSServerV30 struct {
```
Comment (ocket8888) [inline lib/go-tc/deliveryservice_servers.go:109]: Why not make DSServer point to the current definition of a Delivery Service Server? Coupling to the API version means that things stop making sense unless we update that every time the API version changes; like if we have DSServerV30 then APIv3.1 comes out and no changes are made to the structure, we're either using a confusingly version structure, or we have to remember to do something like `type DSServer31 = DSServer30` and then do a find/replace to update things.

File: traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go [hunk 11]
```diff
@@ -554,17 +611,27 @@ func dssSelectQuery() string {
...
-	s.ip_gateway,
-	s.ip_netmask,
+	ARRAY (
+SELECT ( json_build_object (
+'ipAddresses', ARRAY (
+SELECT ( json_build_object (
+'address', ip_address.address,
+'gateway', ip_address.gateway,
+'service_address', ip_address.service_address
+))
```
Comment (ocket8888) [inline traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go:621]: Does `github.com/lib/pq` not support nested arrays?

File: traffic_ops/traffic_ops_golang/deliveryservice/eligible.go [hunk 12]
```diff
@@ -86,13 +105,26 @@ s.ilo_ip_gateway,
...
 s.ilo_username,
-COALESCE(s.interface_mtu, ` + strconv.Itoa(JumboFrameBPS) + `) as interface_mtu,
-s.interface_name,
-s.ip6_address,
-s.ip6_gateway,
-s.ip_address,
-s.ip_gateway,
-s.ip_netmask,
+	ARRAY (
+SELECT ( json_build_object (
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/eligible.go:109]: So this does worry me a bit still. When reviewing @ocket8888's PR I found that this query using `json_build_object` is like 12x slower than querying the interfaces and IPs separately (without using `json_build_object`) and manually populating the servers structs with that data. Also, we should really be making the same query here as we would in the /servers API. It seems like we could make this better by sharing that code/query here.

Locally (which I typically assume performs much better than in Prod), for a DS w/ ~1000 servers assigned, this API takes 850ms for me, whereas before this PR, it takes ~75ms.

Comment (ocket8888) [reply traffic_ops/traffic_ops_golang/deliveryservice/eligible.go:109]: True enough. This is also still how `/servers/details` is being done.

**Pull request merged**